File: src/main/java/com/tkr/wire/SwiftMt940Scanner.java
package com.tkr.wire;

import com.tkr.types.WireTypes.Status;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/** SWIFT MT940 block-4 tag scanner for statement lines and balances. */
public final class SwiftMt940Scanner {

    public enum Mt940Tag {
        TAG_20, TAG_25, TAG_28C, TAG_60F, TAG_60M, TAG_61, TAG_86, TAG_62F, TAG_62M, UNKNOWN
    }

    public static final class Mt940Balance {
        public char dcMark;
        public String date = "";
        public String currency = "";
        public long amountMicro;
        public boolean opening;
    }

    public static final class Mt940StatementLine {
        public String valueDate = "";
        public String entryDate = "";
        public char dcMark;
        public long amountMicro;
        public String transactionType = "";
        public String reference = "";
        public String supplementary = "";
    }

    public static final class Mt940Statement {
        public String transactionRef = "";
        public String accountId = "";
        public String statementNumber = "";
        public Mt940Balance openingBalance;
        public Mt940Balance closingBalance;
        public List<Mt940StatementLine> lines = new ArrayList<>();
        public List<String> infoLines = new ArrayList<>();
    }

    public static final class Mt940ScanResult {
        public Status status = Status.OK;
        public Mt940Statement statement = new Mt940Statement();
        public int tagsParsed;
        public int consumedBytes;
        public int lineCount;
    }

    public static final class Mt940ScannerConfig {
        public boolean strictBlock4 = true;
        public int maxLines = 4096;
        public boolean parseCommaDecimal = true;
    }

    private final Mt940ScannerConfig config;

    public SwiftMt940Scanner() { this(new Mt940ScannerConfig()); }
    public SwiftMt940Scanner(Mt940ScannerConfig config) {
        this.config = config != null ? config : new Mt940ScannerConfig();
    }

    public Mt940ScanResult scanBlock4(byte[] data, int offset, int length) {
        Mt940ScanResult result = new Mt940ScanResult();
        if (data == null || length <= 0) { result.status = Status.TRUNCATED; return result; }
        String text = new String(data, offset, length, StandardCharsets.US_ASCII);
        int blockStart = text.indexOf("{4:");
        if (blockStart >= 0) {
            int blockEnd = text.indexOf("-}", blockStart);
            text = blockEnd > blockStart ? text.substring(blockStart + 3, blockEnd) : text.substring(blockStart + 3);
        }
        List<TagSlice> tags = splitTags(text);
        if (tags.size() > config.maxLines) { result.status = Status.BOUNDS_ERROR; return result; }
        Mt940Statement stmt = new Mt940Statement();
        for (TagSlice tag : tags) {
            Mt940Tag kind = classifyTag(tag.name);
            result.tagsParsed++;
            switch (kind) {
                case TAG_20 -> stmt.transactionRef = tag.body.trim();
                case TAG_25 -> stmt.accountId = tag.body.trim();
                case TAG_28C -> stmt.statementNumber = tag.body.trim();
                case TAG_60F, TAG_60M -> {
                    Mt940Balance bal = parseBalance(tag.body);
                    bal.opening = true;
                    stmt.openingBalance = bal;
                }
                case TAG_62F, TAG_62M -> {
                    Mt940Balance bal = parseBalance(tag.body);
                    bal.opening = false;
                    stmt.closingBalance = bal;
                }
                case TAG_61 -> stmt.lines.add(parseStatementLine(tag.body));
                case TAG_86 -> stmt.infoLines.add(tag.body.trim());
                default -> { }
            }
        }
        result.statement = stmt;
        return result;
    }

    public Mt940ScanResult scanBlock4(byte[] data) {
        return scanBlock4(data, 0, data != null ? data.length : 0);
    }

    public Mt940ScanResult scanBlock4(String text) {
        return scanBlock4(text.getBytes(StandardCharsets.US_ASCII));
    }

    private static final class TagSlice { String name; String body; }

    private List<TagSlice> splitTags(String text) {
        List<TagSlice> out = new ArrayList<>();
        int i = 0;
        while (i < text.length()) {
            if (text.charAt(i) != ':') { i++; continue; }
            int nameEnd = text.indexOf(':', i + 1);
            if (nameEnd < 0) break;
            String name = text.substring(i, nameEnd + 1);
            int nextTag = text.indexOf("\r\n:", nameEnd + 1);
            if (nextTag < 0) nextTag = text.indexOf("\n:", nameEnd + 1);
            String body = nextTag < 0 ? text.substring(nameEnd + 1) : text.substring(nameEnd + 1, nextTag);
            TagSlice slice = new TagSlice();
            slice.name = name;
            slice.body = body;
            out.add(slice);
            i = nextTag < 0 ? text.length() : nextTag + (text.charAt(nextTag) == '\r' ? 2 : 1);
        }
        return out;
    }

    private Mt940Tag classifyTag(String name) {
        if (name == null) return Mt940Tag.UNKNOWN;
        return switch (name) {
            case ":20:" -> Mt940Tag.TAG_20;
            case ":25:" -> Mt940Tag.TAG_25;
            case ":28C:" -> Mt940Tag.TAG_28C;
            case ":60F:" -> Mt940Tag.TAG_60F;
            case ":60M:" -> Mt940Tag.TAG_60M;
            case ":61:" -> Mt940Tag.TAG_61;
            case ":86:" -> Mt940Tag.TAG_86;
            case ":62F:" -> Mt940Tag.TAG_62F;
            case ":62M:" -> Mt940Tag.TAG_62M;
            default -> Mt940Tag.UNKNOWN;
        };
    }

    public Mt940Balance parseBalance(String body) {
        Mt940Balance bal = new Mt940Balance();
        if (body == null || body.length() < 10) return bal;
        bal.dcMark = body.charAt(0);
        bal.date = body.substring(1, 7);
        bal.currency = body.substring(7, 10);
        String amountPart = body.substring(10).replace(",", ".");
        bal.amountMicro = parseAmountMicro(amountPart);
        return bal;
    }

    public Mt940StatementLine parseStatementLine(String body) {
        Mt940StatementLine line = new Mt940StatementLine();
        if (body == null || body.length() < 6) return line;
        line.valueDate = body.substring(0, 6);
        int idx = 6;
        if (body.length() > 10 && Character.isDigit(body.charAt(10))) {
            line.entryDate = body.substring(6, 10);
            idx = 10;
        }
        line.dcMark = body.charAt(idx);
        idx++;
        int refStart = body.indexOf('N', idx);
        if (refStart < 0) refStart = body.indexOf('F', idx);
        String amtStr = refStart > idx ? body.substring(idx, refStart) : body.substring(idx);
        line.amountMicro = parseAmountMicro(amtStr.replace(",", "."));
        if (refStart >= 0 && refStart + 3 < body.length()) {
            line.transactionType = body.substring(refStart, refStart + 3);
            line.reference = body.substring(refStart + 3).trim();
        }
        return line;
    }

    public long parseAmountMicro(String amount) {
        if (amount == null || amount.isEmpty()) return 0;
        boolean neg = amount.endsWith("D") || amount.startsWith("-");
        String cleaned = amount.replace("D", "").replace("C", "").replace("-", "").trim();
        long whole = 0, frac = 0;
        int dot = cleaned.indexOf('.');
        if (dot < 0) {
            whole = parseLongSafe(cleaned);
        } else {
            whole = parseLongSafe(cleaned.substring(0, dot));
            String f = cleaned.substring(dot + 1);
            frac = parseLongSafe(f);
            for (int i = f.length(); i < 6; i++) frac *= 10;
        }
        long micro = whole * 1_000_000L + frac;
        return neg ? -micro : micro;
    }

    private long parseLongSafe(String s) {
        try { return Long.parseLong(s.isEmpty() ? "0" : s); } catch (NumberFormatException e) { return 0; }
    }

    public long sumStatementLines(Mt940Statement stmt) {
        long sum = 0;
        for (Mt940StatementLine line : stmt.lines) {
            sum += line.dcMark == 'D' ? -line.amountMicro : line.amountMicro;
        }
        return sum;
    }

    public boolean reconcileBalances(Mt940Statement stmt) {
        if (stmt.openingBalance == null || stmt.closingBalance == null) return false;
        long open = stmt.openingBalance.dcMark == 'D' ? -stmt.openingBalance.amountMicro : stmt.openingBalance.amountMicro;
        long close = stmt.closingBalance.dcMark == 'D' ? -stmt.closingBalance.amountMicro : stmt.closingBalance.amountMicro;
        return open + sumStatementLines(stmt) == close;
    }
// --- expanded helpers ---
    public int countTags(Mt940ScanResult result) { return result.tagsParsed; }

    public String formatStatementSummary(Mt940Statement stmt) {
        return stmt.transactionRef + "|" + stmt.accountId + "|lines=" + stmt.lines.size();
    }

    public static final class Mt940InfoSegment {
        public String rawCode = "";
        public String narrative = "";
        public List<String> subfields = new ArrayList<>();
    }

    public Mt940ScanResult scan(String swiftText) {
        Mt940ScanResult result = new Mt940ScanResult();
        result.status = Status.OK;
        if (swiftText == null || swiftText.isEmpty()) {
            result.status = Status.TRUNCATED;
            return result;
        }
        int block4Start = swiftText.indexOf("{4:");
        String text = swiftText;
        if (block4Start >= 0) {
            int contentStart = block4Start + 3;
            int block4End = swiftText.indexOf("-}", contentStart);
            if (block4End < 0) {
                result.status = Status.TRUNCATED;
                return result;
            }
            text = swiftText.substring(contentStart, block4End);
            result.consumedBytes = block4End + 2;
        } else {
            result.consumedBytes = swiftText.length();
        }
        result.status = scanBlock4(text, result.statement);
        result.lineCount = result.statement.lines.size();
        result.tagsParsed = fieldsSeen;
        if (result.consumedBytes == 0) result.consumedBytes = swiftText.length();
        return result;
    }

    private int fieldsSeen;

    public Status scanBlock4(String block4, Mt940Statement out) {
        fieldsSeen = 0;
        int cursor = 0;
        while (cursor < block4.length()) {
            char ch = block4.charAt(cursor);
            if (ch == '-' || ch == '\r' || ch == '\n') {
                cursor++;
                continue;
            }
            if (ch != ':') {
                cursor++;
                continue;
            }
            int tagEnd = block4.indexOf(':', cursor + 1);
            if (tagEnd < 0) break;
            String tag = block4.substring(cursor, tagEnd + 1);
            cursor = tagEnd + 1;
            int valueStart = cursor;
            int valueEnd = block4.length();
            int nextTag = block4.indexOf("\n:", cursor);
            if (nextTag >= 0) valueEnd = nextTag;
            String value = trimSwift(block4.substring(valueStart, valueEnd));
            Status st = scanTaggedField(tag, value, out);
            if (st != Status.OK) return st;
            cursor = valueEnd;
        }
        return Status.OK;
    }

    public Status scanTaggedField(String tag, String value, Mt940Statement out) {
        return switch (tag) {
            case ":20:" -> {
                out.transactionRef = value;
                fieldsSeen++;
                yield Status.OK;
            }
            case ":25:" -> {
                out.accountId = value;
                fieldsSeen++;
                yield Status.OK;
            }
            case ":28C:", ":28:" -> {
                out.statementNumber = value;
                fieldsSeen++;
                yield Status.OK;
            }
            case ":60F:", ":60M:" -> {
                fieldsSeen++;
                out.openingBalance = parseBalanceField(value);
                out.openingBalance.opening = true;
                yield Status.OK;
            }
            case ":62F:", ":62M:" -> {
                fieldsSeen++;
                out.closingBalance = parseBalanceField(value);
                out.closingBalance.opening = false;
                yield Status.OK;
            }
            case ":61:" -> {
                Mt940StatementLine line = parseStatementLineField(value);
                out.lines.add(line);
                fieldsSeen++;
                yield Status.OK;
            }
            case ":86:" -> {
                out.infoLines.add(parseField86(value));
                fieldsSeen++;
                yield Status.OK;
            }
            default -> Status.OK;
        };
    }

    private Mt940Balance parseBalanceField(String body) {
        Mt940Balance bal = parseBalance(body);
        return bal;
    }

    private String parseField86(String body) {
        Mt940InfoSegment seg = new Mt940InfoSegment();
        scanField86(body, seg);
        return seg.narrative.isEmpty() ? body.trim() : seg.narrative;
    }

    public Status scanField86(String fieldBody, Mt940InfoSegment out) {
        out.rawCode = "";
        out.narrative = "";
        out.subfields.clear();
        String body = trimSwift(fieldBody);
        if (body.isEmpty()) return Status.TRUNCATED;
        if (body.length() >= 3 && body.charAt(0) == '/' && Character.isLetter(body.charAt(1))) {
            int endCode = body.indexOf('/', 2);
            if (endCode >= 0) {
                out.rawCode = body.substring(0, endCode);
                body = body.substring(endCode);
            }
        }
        splitSubfields86(body, out.subfields);
        if (!out.subfields.isEmpty()) out.narrative = out.subfields.get(out.subfields.size() - 1);
        else out.narrative = body;
        return Status.OK;
    }

    public Status splitSubfields86(String body, List<String> subfields) {
        subfields.clear();
        int cursor = 0;
        while (cursor < body.length()) {
            if (cursor + 3 <= body.length() && body.charAt(cursor) == '/'
                    && Character.isLetter(body.charAt(cursor + 1))) {
                int next = body.indexOf('/', cursor + 2);
                if (next < 0) {
                    subfields.add(body.substring(cursor));
                    break;
                }
                subfields.add(body.substring(cursor, next));
                cursor = next;
                continue;
            }
            int nl = body.indexOf('\n', cursor);
            if (nl < 0) {
                subfields.add(body.substring(cursor));
                break;
            }
            subfields.add(body.substring(cursor, nl));
            cursor = nl + 1;
        }
        return Status.OK;
    }

    public Mt940StatementLine parseStatementLineField(String fieldBody) {
        Mt940StatementLine out = new Mt940StatementLine();
        if (fieldBody == null || fieldBody.length() < 12) return out;
        int cursor = 0;
        out.valueDate = fieldBody.substring(cursor, cursor + 6);
        cursor += 6;
        if (cursor + 4 <= fieldBody.length() && Character.isDigit(fieldBody.charAt(cursor))) {
            out.entryDate = fieldBody.substring(cursor, cursor + 4);
            cursor += 4;
        }
        out.dcMark = fieldBody.charAt(cursor++);
        if (cursor < fieldBody.length() && "CDRS".indexOf(fieldBody.charAt(cursor)) >= 0) cursor++;
        int amountStart = cursor;
        while (cursor < fieldBody.length() && (Character.isDigit(fieldBody.charAt(cursor)) || fieldBody.charAt(cursor) == ',')) {
            cursor++;
        }
        out.amountMicro = parseAmountMicro(fieldBody.substring(amountStart, cursor).replace(",", "."));
        if (cursor + 4 <= fieldBody.length()) {
            out.transactionType = fieldBody.substring(cursor, cursor + 4);
            cursor += 4;
        }
        if (cursor < fieldBody.length()) {
            String refs = fieldBody.substring(cursor).trim();
            int slash = refs.indexOf("//");
            if (slash >= 0) {
                out.reference = refs.substring(0, slash);
                out.supplementary = refs.substring(slash + 2);
            } else {
                out.reference = refs;
            }
        }
        return out;
    }

    public long parseAmountMinor(String amountText, char debitCredit) {
        long minor = parseAmountMicro(amountText.replace(",", ".")) / 10000L;
        if (debitCredit == 'D' || debitCredit == 'd') minor = -minor;
        return minor;
    }

    private static String trimSwift(String text) {
        int start = 0;
        int end = text.length();
        while (start < end && (text.charAt(start) == '\r' || text.charAt(start) == '\n' || text.charAt(start) == ' ')) start++;
        while (end > start && (text.charAt(end - 1) == '\r' || text.charAt(end - 1) == '\n' || text.charAt(end - 1) == ' ')) end--;
        return text.substring(start, end);
    }

    public static boolean isDebitCredit(char dc) {
        return dc == 'C' || dc == 'D' || dc == 'R' || dc == 'c' || dc == 'd';
    }
}

File: legacy-cpp/src/wire/swift_mt940_scanner.cc
#include "tkr/wire/swift_mt940_scanner.h"

#include "tkr/util/bounds.h"

#include <algorithm>
#include <cctype>

namespace tkr {
namespace wire {
namespace {

bool IsSwiftDigit(char ch) {
  return ch >= '0' && ch <= '9';
}

bool IsSwiftAlpha(char ch) {
  return (ch >= 'A' && ch <= 'Z') || (ch >= 'a' && ch <= 'z');
}

std::string_view TrimSwift(std::string_view text) {
  while (!text.empty() && (text.front() == '\r' || text.front() == '\n' ||
                           text.front() == ' ')) {
    text.remove_prefix(1);
  }
  while (!text.empty() && (text.back() == '\r' || text.back() == '\n' ||
                           text.back() == ' ')) {
    text.remove_suffix(1);
  }
  return text;
}

Status ParseYYMMDD(std::string_view text, std::string* out) {
  if (out == nullptr) {
    return Status::kBoundsError;
  }
  if (text.size() != 6) {
    return Status::kBoundsError;
  }
  for (char ch : text) {
    if (!IsSwiftDigit(ch)) {
      return Status::kBoundsError;
    }
  }
  out->assign(text.data(), text.size());
  return Status::kOk;
}

}  // namespace

SwiftMt940Scanner::SwiftMt940Scanner() : fields_seen_(0) {}

bool SwiftMt940Scanner::IsDebitCredit(char dc) {
  return dc == 'C' || dc == 'D' || dc == 'R' || dc == 'c' || dc == 'd';
}

Status SwiftMt940Scanner::ParseBalance(std::string_view body, Mt940Balance* out) {
  if (out == nullptr || body.size() < 10) {
    return Status::kBoundsError;
  }
  out->debit_credit = body[0];
  if (!IsDebitCredit(out->debit_credit)) {
    return Status::kBoundsError;
  }
  Status date_status = ParseYYMMDD(body.substr(1, 6), &out->date_yymmdd);
  if (date_status != Status::kOk) {
    return date_status;
  }
  const std::size_t curr_start = 7;
  if (!util::SliceInBounds(curr_start, 3, body.size())) {
    return Status::kBoundsError;
  }
  out->currency.assign(body.substr(curr_start, 3));
  std::string_view amount_text = body.substr(curr_start + 3);
  return SwiftMt940Scanner{}.ParseAmountMinor(amount_text, out->debit_credit,
                                            &out->amount_minor);
}

Status SwiftMt940Scanner::ParseAmountMinor(std::string_view amount_text,
                                           char debit_credit,
                                           std::int64_t* out) {
  if (out == nullptr || amount_text.empty()) {
    return Status::kBoundsError;
  }
  std::int64_t whole = 0;
  std::int64_t frac = 0;
  bool seen_comma = false;
  std::size_t frac_digits = 0;
  for (char ch : amount_text) {
    if (ch == ',') {
      seen_comma = true;
      continue;
    }
    if (!IsSwiftDigit(ch)) {
      break;
    }
    const std::int64_t digit = static_cast<std::int64_t>(ch - '0');
    if (!seen_comma) {
      whole = whole * 10 + digit;
    } else if (frac_digits < 2) {
      frac = frac * 10 + digit;
      ++frac_digits;
    }
  }
  while (frac_digits < 2) {
    frac *= 10;
    ++frac_digits;
  }
  std::int64_t minor = whole * 100 + frac;
  if (debit_credit == 'D' || debit_credit == 'd') {
    minor = -minor;
  }
  *out = minor;
  return Status::kOk;
}

Status SwiftMt940Scanner::ScanField61(std::string_view field_body,
                                      Mt940StatementLine* out) {
  if (out == nullptr || field_body.size() < 12) {
    return Status::kBoundsError;
  }
  out->value_date.clear();
  out->entry_date.clear();
  out->supplementary.clear();
  out->transaction_type.clear();
  out->customer_reference.clear();
  out->bank_reference.clear();

  std::size_t cursor = 0;
  Status vd = ParseYYMMDD(field_body.substr(cursor, 6), &out->value_date);
  if (vd != Status::kOk) {
    return vd;
  }
  cursor += 6;

  if (cursor + 4 <= field_body.size() && IsSwiftDigit(field_body[cursor])) {
    Status ed = ParseYYMMDD(field_body.substr(cursor, 4), &out->entry_date);
    if (ed == Status::kOk) {
      cursor += 4;
    }
  }

  if (cursor >= field_body.size()) {
    return Status::kTruncated;
  }
  out->debit_credit = field_body[cursor++];
  if (!IsDebitCredit(out->debit_credit)) {
    return Status::kBoundsError;
  }

  if (cursor < field_body.size() &&
      (field_body[cursor] == 'C' || field_body[cursor] == 'D' ||
       field_body[cursor] == 'R' || field_body[cursor] == 'S')) {
    out->funds_code = field_body[cursor++];
  } else {
    out->funds_code = 'N';
  }

  const std::size_t amount_start = cursor;
  while (cursor < field_body.size() &&
         (IsSwiftDigit(field_body[cursor]) || field_body[cursor] == ',')) {
    ++cursor;
  }
  if (cursor == amount_start) {
    return Status::kBoundsError;
  }
  Status amt = ParseAmountMinor(field_body.substr(amount_start, cursor - amount_start),
                                out->debit_credit, &out->amount_minor);
  if (amt != Status::kOk) {
    return amt;
  }

  if (cursor + 4 <= field_body.size()) {
    out->transaction_type.assign(field_body.substr(cursor, 4));
    cursor += 4;
  }

  const std::size_t ref_start = cursor;
  while (cursor < field_body.size() && field_body[cursor] != '\n' &&
         field_body[cursor] != '\r') {
    ++cursor;
  }
  if (cursor > ref_start) {
    std::string_view refs = field_body.substr(ref_start, cursor - ref_start);
    const std::size_t slash = util::FindSubstring(refs, "//", 0);
    if (slash != std::string_view::npos) {
      out->customer_reference.assign(refs.substr(0, slash));
      out->bank_reference.assign(refs.substr(slash + 2));
    } else {
      out->customer_reference.assign(refs.data(), refs.size());
    }
  }
  return Status::kOk;
}

Status SwiftMt940Scanner::SplitSubfields86(std::string_view body,
                                           std::vector<std::string>* subfields) {
  if (subfields == nullptr) {
    return Status::kBoundsError;
  }
  subfields->clear();
  std::size_t cursor = 0;
  while (cursor < body.size()) {
    if (cursor + 3 <= body.size() && body[cursor] == '/' &&
        IsSwiftAlpha(body[cursor + 1])) {
      const std::size_t next = util::FindSubstring(body, "/", cursor + 2);
      if (next == std::string_view::npos) {
        subfields->emplace_back(body.substr(cursor));
        break;
      }
      subfields->emplace_back(body.substr(cursor, next - cursor));
      cursor = next;
      continue;
    }
    const std::size_t nl = body.find('\n', cursor);
    if (nl == std::string_view::npos) {
      subfields->emplace_back(body.substr(cursor));
      break;
    }
    subfields->emplace_back(body.substr(cursor, nl - cursor));
    cursor = nl + 1;
  }
  return Status::kOk;
}

Status SwiftMt940Scanner::ScanField86(std::string_view field_body,
                                      Mt940InfoSegment* out) {
  if (out == nullptr) {
    return Status::kBoundsError;
  }
  out->raw_code.clear();
  out->narrative.clear();
  out->subfields.clear();

  std::string_view body = TrimSwift(field_body);
  if (body.empty()) {
    return Status::kTruncated;
  }

  if (body.size() >= 3 && body[0] == '/' && IsSwiftAlpha(body[1])) {
    const std::size_t end_code = body.find('/', 2);
    if (end_code != std::string_view::npos) {
      out->raw_code.assign(body.substr(0, end_code));
      body = body.substr(end_code);
    }
  }

  Status split = SplitSubfields86(body, &out->subfields);
  if (split != Status::kOk) {
    return split;
  }
  if (!out->subfields.empty()) {
    out->narrative = out->subfields.back();
  } else {
    out->narrative.assign(body.data(), body.size());
  }
  return Status::kOk;
}

Status SwiftMt940Scanner::ScanTaggedField(std::string_view tag,
                                          std::string_view value,
                                          Mt940Statement* out) {
  if (out == nullptr) {
    return Status::kBoundsError;
  }
  if (tag == ":20:") {
    out->transaction_reference.assign(value.data(), value.size());
    ++fields_seen_;
    return Status::kOk;
  }
  if (tag == ":25:") {
    out->account_id.assign(value.data(), value.size());
    ++fields_seen_;
    return Status::kOk;
  }
  if (tag == ":28C:" || tag == ":28:") {
    out->statement_number.assign(value.data(), value.size());
    ++fields_seen_;
    return Status::kOk;
  }
  if (tag == ":60F:" || tag == ":60M:") {
    ++fields_seen_;
    return ParseBalance(value, &out->opening);
  }
  if (tag == ":62F:" || tag == ":62M:") {
    ++fields_seen_;
    return ParseBalance(value, &out->closing);
  }
  if (tag == ":61:") {
    Mt940StatementLine line;
    Status st = ScanField61(value, &line);
    if (st == Status::kOk) {
      out->lines.push_back(line);
      ++fields_seen_;
    }
    return st;
  }
  if (tag == ":86:") {
    Mt940InfoSegment info;
    Status st = ScanField86(value, &info);
    if (st == Status::kOk) {
      out->info_segments.push_back(info);
      ++fields_seen_;
    }
    return st;
  }
  return Status::kOk;
}

Status SwiftMt940Scanner::ScanBlock4(std::string_view block4, Mt940Statement* out) {
  if (out == nullptr) {
    return Status::kBoundsError;
  }
  std::size_t cursor = 0;
  while (cursor < block4.size()) {
    if (block4[cursor] == '-' || block4[cursor] == '\r' ||
        block4[cursor] == '\n') {
      ++cursor;
      continue;
    }
    if (block4[cursor] != ':') {
      ++cursor;
      continue;
    }
    const std::size_t tag_end = block4.find(':', cursor + 1);
    if (tag_end == std::string_view::npos) {
      break;
    }
    std::string_view tag = block4.substr(cursor, tag_end - cursor + 1);
    cursor = tag_end + 1;
    const std::size_t value_start = cursor;
    std::size_t value_end = block4.size();
    const std::size_t next_tag = block4.find("\n:", cursor);
    if (next_tag != std::string_view::npos) {
      value_end = next_tag;
    }
    if (!util::SectionBodyInBounds(value_start, value_end - value_start,
                                   block4.size())) {
      return Status::kBoundsError;
    }
    std::string_view value = TrimSwift(block4.substr(value_start, value_end - value_start));
    Status st = ScanTaggedField(tag, value, out);
    if (st != Status::kOk) {
      return st;
    }
    cursor = value_end;
  }
  return Status::kOk;
}

Mt940ScanResult SwiftMt940Scanner::Scan(std::string_view swift_text) {
  Mt940ScanResult result{};
  result.status = Status::kOk;
  result.consumed_bytes = 0;
  result.line_count = 0;
  fields_seen_ = 0;

  std::string_view text = swift_text;
  const std::size_t block4_start = util::FindSubstring(text, "{4:", 0);
  if (block4_start != std::string_view::npos) {
    const std::size_t content_start = block4_start + 3;
    const std::size_t block4_end = text.find("-}", content_start);
    if (block4_end == std::string_view::npos) {
      result.status = Status::kTruncated;
      return result;
    }
    if (!util::SectionBodyInBounds(content_start, block4_end - content_start,
                                   text.size())) {
      result.status = Status::kBoundsError;
      return result;
    }
    text = text.substr(content_start, block4_end - content_start);
    result.consumed_bytes = block4_end + 2;
  } else {
    result.consumed_bytes = text.size();
  }

  result.status = ScanBlock4(text, &result.statement);
  result.line_count = static_cast<std::uint32_t>(result.statement.lines.size());
  if (result.consumed_bytes == 0) {
    result.consumed_bytes = swift_text.size();
  }
  return result;
}

}  // namespace wire
}  // namespace tkr
